### ISSUE TITLE
User binds input no longer display retro_input.cpp

### ISSUE THREAD

**Issue by RagingNoob** [libretro/fbalpha#213]
This issue has been randomly appearing throughout the commit history.
I thought it was the upgrade to 1.7.6 but its the FBAlpha core.

[https://i.ibb.co/18Nh2B9/RA-input-comparison.png](https://i.ibb.co/18Nh2B9/RA-input-comparison.png)

As the image demonstrates

**User 1 Weak Kick** has been inadvertently replaced by 
**User 1 B button (down)**

etc.
The last commit I have tested that works correctly is  **#5d60621** Mar 19, 2019

**Comment by barbudreadmon** [comment]
I saw the ping from @fr500 on discord.

![img2](https://user-images.githubusercontent.com/10391265/56015918-debcb800-5cfa-11e9-93a2-ffe97f6b86eb.png)
As you can see in this screenshot from `Settings > Input > User 1 Binds`, right half is the same as you, while left half shows that i'm sending input descriptors with the correct labels, furthermore if you go in `Quick Menu > Controls` instead, which is the right place for core/game remapping, you'll get this : 
![img1](https://user-images.githubusercontent.com/10391265/56016176-ba151000-5cfb-11e9-8a4a-bd3db32fe26f.png)
So, labels like "Weak Punch" are sent properly to RA since they are showing in this menu.

So : 
- I'm sending input descriptors with the right labels, and it's obviously working properly because those labels are showing in the menu dedicated to this
- I don't understand why you are using this menu in the first place, `Settings > Input > User 1 binds` is about establishing the relationship between the actual controller and the retropad layer, while `Quick menu > Controls` is about establishing relationship between the retropad layer and the core buttons, but anyway if RA wants to show the labels in both, that's its decision, i've no hand in that, my job is to send input descriptors through the api call, which i do, and it's obviously working as expected

At this point, while i totally admit a change in the core code is the root of your "issue", i can't consider it a core issue, the frontend is the one deciding what it does with the information i send, if it wants to show the labels i send in `Settings > Input > User 1 binds`, so be it.

About the change that might have caused this, i've a pretty good guess, until recently i was doing a lot of unnecessary calls to `RETRO_ENVIRONMENT_SET_INPUT_DESCRIPTORS` with the same data, while it should be called only when there is something new to say about the input descriptors. I trimmed down those duplicate calls, i've no intention to bring them back.

For all those reasons, i'll tag this as wontfix.

**Comment by barbudreadmon** [comment]
Oh, and btw, out of good faith, i digged the cause of this "issue", and uncommenting https://github.com/libretro/fbalpha/blob/0be3e15495fe5d87912fd874540d39b0e5c284be/src/burner/libretro/retro_input.cpp#L1890 fix it. It's an early call triggered in `retro_load_game`, while i now wait for `retro_set_controller_port_device` being called for each device before setting this, imho there is point in calling this before we even know what device (Classic, Modern, Mouse ball, Mouse full, Pointer or Lightgun) is connected.

I think the new behavior is the right one, and the old behavior was messy : having labels in this menu prevent proper controller to retropad mapping, because it's supposed to be done while abstracting the core button layer, unlike the other menu.
If @fr500 tells me the previous behavior was the right one, then so be it, i'll change back to the old behavior, even if i think this decision should be handled frontend-side, until then i'll stick with the new one

**Comment by RagingNoob** [comment]
Thank you for taking the time to look at this.

**Settings > Input > User 1 Binds**

This has always been the method that I use to configure inputs for RA.
 I was never told to not use those options.

There are a lot of popular fighting style games on FBAlpha and subsequently a lot of 6/8 button fightsticks layed out in a typical 2 row (3x2) or (4x2) horizontal array.
Some sticks I have tested do not have YXBA or square circle triangle printed on them etc so it was far easier to have them replaced by the GUI as Weak Punch, Heavy Kick etc.

The **User 1 Device Type : Retropad, Classic, Modern** is confusing as changing them to any of those 3 doesn't appear to make any difference.

Instead you are saying I should only use the **Quick Menu > Controls** 

I have tried this method but it is way to cumbersome, ineffecient and even more confusing.
I have always been active to encourage others to use Retroarch but with the recent FBAlpha core versions the joystick/fightstick configuration is a headache - a discouragement - when it could easily be straightforward and intuitive to the end user to complete the input config without any intervention or advice.

The overall ability to have universal controls in FBAlpha/Retroarch was to configure once and not have to reconfigure again for every single game, but setting it up in the first place is a real challenge!

**Comment by barbudreadmon** [comment]
So, that's the order RA is calling things : 
1. `retro_load_game` <= load the game
2. `retro_set_controller_port_device` <= called for each device so that the core knows what is connected
3. `retro_run` <= looped, rendering 1 frame each time

Apparently, the labels will show in `Settings > Input > User 1 Binds` only if the input descriptors were sent in `retro_load_game`, which is a complete non-sense : **input descriptors are dependent on what kind of device is connected, which is something RA is telling to the core later**.
I'm glad we are having this discussion, because this stuff needs clarification.

But anyway, the main issue i see here is that the way you are dealing with bindings is wrong, RA don't care if your controller has no YXBA label, the main concept of RA input system is to have this retropad abstraction layer between the core and the actual controller, however you are trying to bypass it, this is the wrong way to deal with this inputs in RA. Your job if RA isn't recognizing your controller instantly is to properly map it to the retropad model in the `Settings > Input > User 1 Binds` menu.

To follow the retropad model with fightsticks/panels, it is easy, 6 buttons must always be mapped like this : 
`Y X L`
`B A R`
8 buttons must be mapped like this : 
`Y X R1 L1`
`B A R2 L2`
NB : some people will swap columns and map 8 buttons like this : 
`L1 Y X R1`
`L2 B A R2`
It's acceptable too, and might even be recommended if you do weird things like playing Mario Kart with fightsticks/panels (so that L are left and R are right).

As soon as you did this first step properly, managing things in `Quick menu > Controls` will be easy, there are explanation about what is Classic and Modern at https://github.com/libretro/fbalpha , so you just have to select the one you see fit. FBAlpha default mappings are done so that they can work out-of-the-box, so simply setting one of those 2 value will be ok for 99.9% of people and games, but you can further customize mapping and use "save game remap file" in this menu if it doesn't fully suit your taste.

**Comment by furiadeoso** [comment]
@RagingNoob As far as I know the correct way to use controls is:
First: Be sure out of any game that your real control is binding to retropad. Knowing this binding.


Second: If needed, remap controls for a core and/or a game. 
In this second moment you reassign the default retropad buttons to other actions (punch, kick, or whatever).


I think that following this method defining controls is fast and very easy. Try first to find a suitable controls remap for the whole core and after make exceptions per game.

**Comment by barbudreadmon** [comment]
@furiadeoso thanks for explaining in simpler terms, my explanation might be overwhelming
